### PR TITLE
Add basic support for colored printers

### PR DESCRIPTION
**Pull request by nicolasstucki**
With this change, both `dotc -decompile` and `dotc -Xprint` get colored outputs.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/printing/RefinedPrinter.scala [hunk 1]
```diff
@@ -346,29 +346,31 @@ class RefinedPrinter(_ctx: Context) extends PlainPrinter(_ctx) {
       case id: Trees.SearchFailureIdent[_] =>
         tree.typeOpt match {
           case reason: Implicits.SearchFailureType =>
-            toText(id.name) ~ "implicitly[" ~ toText(reason.expectedType) ~ "]"
```
Comment (OlivierBlanvillain) [inline compiler/src/dotty/tools/dotc/printing/RefinedPrinter.scala]: A `[` disappeared, not sure if it's intentional

File: compiler/src/dotty/tools/dotc/printing/RefinedPrinter.scala [hunk 2]
```diff
@@ -400,25 +402,25 @@ class RefinedPrinter(_ctx: Context) extends PlainPrinter(_ctx) {
...
       case Closure(env, ref, target) =>
         "closure(" ~ (toTextGlobal(env, ", ") ~ " | " provided env.nonEmpty) ~
         toTextGlobal(ref) ~ (":" ~ toText(target) provided !target.isEmpty) ~ ")"
       case Match(sel, cases) =>
         if (sel.isEmpty) blockText(cases)
-        else changePrec(GlobalPrec) { toText(sel) ~ " match " ~ blockText(cases) }
+        else changePrec(GlobalPrec) { toText(sel) ~ keywordStr(" match ") ~ blockText(cases) }
       case CaseDef(pat, guard, body) =>
-        "case " ~ inPattern(toText(pat)) ~ optText(guard)(" if " ~ _) ~ " => " ~ caseBlockText(body)
+        keywordStr("case ") ~ inPattern(toText(pat)) ~ optText(guard)(" if " ~ _) ~ " => " ~ caseBlockText(body)
```
Comment (OlivierBlanvillain) [inline compiler/src/dotty/tools/dotc/printing/RefinedPrinter.scala]: this `if` is not highlighted

File: compiler/src/dotty/tools/dotc/printing/SyntaxHighlighting.scala [hunk 3]
```diff
@@ -17,6 +17,7 @@ object SyntaxHighlighting {
   val KeywordColor    = Console.YELLOW
   val ValDefColor     = Console.CYAN
   val LiteralColor    = Console.RED
+  val StringColor    = Console.GREEN
```
Comment (OlivierBlanvillain) [inline compiler/src/dotty/tools/dotc/printing/SyntaxHighlighting.scala]: Missing space ^^'

File: compiler/src/dotty/tools/dotc/util/DiffUtil.scala [hunk 4]
```diff
@@ -5,31 +5,27 @@ import scala.collection.mutable
...
-        val (token, rest) = if (Character.isAlphabetic(head) || Character.isDigit(head)) {
-          str.span(c => Character.isAlphabetic(c) || Character.isDigit(c))
-        } else if (Character.isMirrored(head) || Character.isWhitespace(head)) {
-          str.splitAt(1)
-        } else {
-          str.span { c =>
-            !Character.isAlphabetic(c) && !Character.isDigit(c) &&
-              !Character.isMirrored(c) && !Character.isWhitespace(c)
+        val (token, rest) =
+          if (head == '\u001b') { // ansi color token
```
Comment (OlivierBlanvillain) [inline compiler/src/dotty/tools/dotc/util/DiffUtil.scala]:
```scala
val ansiColorToken = '\u001b'
```

(also to be used below)

File: compiler/src/dotty/tools/dotc/util/DiffUtil.scala [hunk 5]
```diff
@@ -69,25 +65,21 @@ object DiffUtil {
     }
 
     val expectedDiff =
-      if (expected eq EOF) ANSI_EOF + expected + ANSI_DEFAULT
+      if (expected eq EOF) eof()
```
Comment (OlivierBlanvillain) [inline compiler/src/dotty/tools/dotc/util/DiffUtil.scala]: You are printing less here (`expected` is not shown), is it intentional?

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/util/DiffUtil.scala]: The string for expected is in this case `"EOF"`. `EOF` is a string that has `"EOF"` but is also meant to be unique.

File: compiler/src/dotty/tools/dotc/printing/SyntaxHighlighting.scala (hunk 3 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/printing/SyntaxHighlighting.scala]: Nooooo. Shame on me.

**Pull request merged**